pkg/image/api/types.go

```diff
@@ -60,6 +77,41 @@ type ImageLayer struct {
 	Size int64
 }
 
+// SignatureValidationResult holds a result of signature validation.
+type SignatureValidationResult string
+
+// ImageSignature holds a signature of an image.
+type ImageSignature struct {
+	// Describes a type of stored blob. Provided by caller.
+	Type string
+	// An opaque binary string which is an image's signature.
+	Content []byte
+	// A time of last server-side validation.
+	ValidatedOn unversioned.Time
+	// Result of validation performed at ValidatedOn.
+	ValidationResult SignatureValidationResult
+
+	// At most one of the following pointers can point to details. If the signature type isn't recognized,
+	// they will all be nil.
+
+	// Details for signature of type "atomic".
+	AtomicSignatureDetails *AtomicSignatureDetails
+	// For any future supported signature type, add additional pointer to its own details struct.
+}
+
+// AtomicSignatureDetails holds metadata of recognized signature of type "atomic". They are parsed from
+// signature's content.
+type AtomicSignatureDetails struct {
+	// Digest of signed manifest.
+	SignedImageName string
+	// Docker image reference
+	SignedDockerImageReference string
```

Comment:
These references are almost never going to match real values in a
production system.  For one thing, we want images to be pullable from
multiple sources.  It's useful to attest to the name, but the digest is the
value that we should be verifying against.

On Mon, May 23, 2016 at 5:21 PM, Miloslav Trmač <notifications@github.com>
wrote:

> In pkg/image/api/types.go
> <https://github.com/openshift/origin/pull/8371#discussion_r64291416>:
>
> > +
> > +	// At most one of the following pointers can point to details. If the signature type isn't recognized,
> > +	// they will all be nil.
> > +
> > +	// Details for signature of type "atomic".
> > +	AtomicSignatureDetails *AtomicSignatureDetails
> > +	// For any future supported signature type, add additional pointer to its own details struct.
> > +}
> > +
> > +// AtomicSignatureDetails holds metadata of recognized signature of type "atomic". They are parsed from
> > +// signature's content.
> > +type AtomicSignatureDetails struct {
> > +	// Digest of signed manifest.
> > +	SignedImageName string
> > +	// Docker image reference
> > +	SignedDockerImageReference string
>
> Do you expect to verify this Name against something?
>
> Yes, definitely.
>
> The simplest case is verification when pulling from an external/upstream
> registry, in which case the signed identity should match the name we are
> pulling (e.g. when pulling registry.access.redhat.com/rhel7/rhel:$sometag,
> the signature should designate the same identity).
>
> The policy evaluation engine (WIP
> https://github.com/projectatomic/skopeo/pull/55/files ) also allows
> overriding the verified identity. So, a hypothetical OpenShift imagestream
> containing locally approved RHEL base images could be configured to require
> two signatures, one by RH using the RHEL identity, and one from the local
> team responsible for the approval, using the imagestream identity.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8371/files/a49f7a114201f3d8868dba2c837f84b85045e79c#r64291416>
>